rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java

```diff
@@ -372,11 +289,233 @@ private void handleException(RabbitListenerException exception) {
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } finally {
+                consumer = null;
+                consumers.remove(this);
+                channelPool.returnChannel(channel);
+            }
+        }
+
+        private void performConsumerRecovery() {
+            com.rabbitmq.client.DefaultConsumer recoveredConsumer = null;
+            int recoveryAttempts = 0;
+            while (recoveredConsumer == null) {
+                Channel channel = null;
+                try {
+                    synchronized (this) {
+                        if (canceled) {
+                            return;
+                        }
+                        LOG.debug("consumer [{}] recovery attempt: {}", consumerTag, recoveryAttempts + 1);
+                        channel = channelPool.getChannelWithRecoveringDelay(recoveryAttempts++);
+                        recoveredConsumer = createConsumer(channel);
+                        consumer = recoveredConsumer;
+                    }
+                } catch (IOException e) {
+                    if (channel != null) {
+                        channelPool.returnChannel(channel);
+                    }
+                    LOG.warn("Recovery attempt {} for consumer [{}] failed, will retry.",
+                            recoveryAttempts, consumerTag, e);
+                } catch (InterruptedException e) {
+                    LOG.warn("The consumer [{}] recovery was interrupted. The consumer will not recover.",
+                            consumerTag, e);
+                    Thread.currentThread().interrupt();
+                    return;
+                }
+            }
+            LOG.info("consumer [{}] recovered", consumerTag);
+        }
+
+        private com.rabbitmq.client.DefaultConsumer createConsumer(Channel channel) throws IOException {
+            setChannelPrefetch(prefetch, channel);
+
+            com.rabbitmq.client.DefaultConsumer consumer = new com.rabbitmq.client.DefaultConsumer(channel) {
+                /**
+                 * The consumer was irregular terminated. This may be caused by a deleted or temporary unavailable
+                 * queue.
+                 * <p>
+                 * This kind of infrastructure failure may happen due to RabbitMQ cluster node restarts or other
+                 * external actions. The client application will most likely be unable to restore the infrastructure,
+                 * but it should return to normal operation as soon as the external infrastructure problem is solved.
+                 * e.g. the RabbitMQ node restart is complete and the queue is available again.
+                 */
+                @Override
+                public void handleCancel(String consumerTag) throws IOException {
+                    try {
+                        getChannel().close(AMQP.REPLY_SUCCESS, "closing channel of canceled consumer");
```

Comment:
Can you explain why the channel needs to be closed if the consumer was cancelled?

Reply:
The thought was like "if the consumer dies, its channel has to die also", but you are right. The channel can be reused.